File: frontend/types/product.ts
export type ProductRead = {
    id: number
    system_id: string
    sku: string
    name: string
    description: string | null
    usage: string | null
    expiration: string | null
    priority: number
    in_stock: boolean
    archived: boolean
    rating_avg: number
    rating_count: number
    image_url: string
    created_at: string
    updated_at: string
}

export type ProductVariantRead = {
    id: number
    system_id: string
    image_url: string
    sku: string | null
    name: string
    stock: number
    archived: boolean
    price: string
    created_at: string
    updated_at: string
}

export type ProductWithVariantsRead = ProductRead & {
    variants: ProductVariantRead[]
}

export type ProductReviewRead = {
    id: number
    author_username: string
    product_id: number
    value: number
    text: string | null
    answer: string | null
    attachments: ProductReviewAttachmentRead[]
    likes: number
    dislikes: number
    moderated: boolean
    created_at: string
    updated_at: string
}

export type ProductReviewAttachmentRead = {
    id: number
    image_url: string
    created_at: string
    updated_at: string
}

export type UploadableReviewAttachment = {
    uri: string
    fileName?: string | null
    mimeType?: string | null
}

export type ProductReviewCreate = {
    value: number
    text?: string | null
    attachments?: UploadableReviewAttachment[]
}

export type ProductReviewEligibilityRead = {
    can_review: boolean
}

export type ProductCreate = {
    system_id?: string | null
    sku: string
    name: string
    description?: string | null
    usage?: string | null
    expiration?: string | null
    archived?: boolean
    priority?: number
}

export type ProductUpdate = {
    system_id?: string | null
    sku?: string | null
    name?: string | null
    description?: string | null
    usage?: string | null
    expiration?: string | null
    archived?: boolean | null
    priority?: number | null
}

File: backend/src/database/schemas/catalog/product.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator

from src.database.limits import (
    PRODUCT_DESCRIPTION_MAX_LENGTH,
    PRODUCT_EXPIRATION_MAX_LENGTH,
    PRODUCT_NAME_MAX_LENGTH,
    PRODUCT_SKU_MAX_LENGTH,
    PRODUCT_USAGE_MAX_LENGTH,
)
from src.database.product_text import normalize_product_text
from src.database.schemas.catalog.variant import ProductVariantRead


class ProductBase(BaseModel):
    sku: str = Field(min_length=1, max_length=PRODUCT_SKU_MAX_LENGTH)
    name: str = Field(min_length=1, max_length=PRODUCT_NAME_MAX_LENGTH)
    description: str | None = Field(default=None, max_length=PRODUCT_DESCRIPTION_MAX_LENGTH)
    usage: str | None = Field(default=None, max_length=PRODUCT_USAGE_MAX_LENGTH)
    expiration: str | None = Field(default=None, max_length=PRODUCT_EXPIRATION_MAX_LENGTH)
    priority: int = Field(default=0, ge=0)

    @field_validator("description", "usage", "expiration", mode="before")
    @classmethod
    def normalize_rich_text(cls, value):
        return normalize_product_text(value)


class ProductCreate(ProductBase):
    system_id: uuid.UUID | None = None


class ProductUpdate(BaseModel):
    sku: str | None = Field(default=None, min_length=1, max_length=PRODUCT_SKU_MAX_LENGTH)
    name: str | None = Field(default=None, min_length=1, max_length=PRODUCT_NAME_MAX_LENGTH)
    description: str | None = Field(default=None, max_length=PRODUCT_DESCRIPTION_MAX_LENGTH)
    usage: str | None = Field(default=None, max_length=PRODUCT_USAGE_MAX_LENGTH)
    expiration: str | None = Field(default=None, max_length=PRODUCT_EXPIRATION_MAX_LENGTH)
    priority: int | None = Field(default=None, ge=0)
    system_id: uuid.UUID | None = None

    @field_validator("description", "usage", "expiration", mode="before")
    @classmethod
    def normalize_rich_text(cls, value):
        return normalize_product_text(value)


class ProductRead(ProductBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    system_id: uuid.UUID
    in_stock: bool
    image_url: str
    rating_avg: float = 0.0
    rating_count: int = 0
    created_at: datetime
    updated_at: datetime


class ProductWithVariantsRead(ProductRead):
    variants: list[ProductVariantRead]
